### PR TITLE
TM-1037 Connect integration with user

### PR DESCRIPTION
**Pull request by bua89**

### CODE REVIEW

File: src/SevenTag/Api/SecurityBundle/DataFixtures/ORM/IntegrationFixture.php [hunk 1]
```diff
@@ -37,6 +38,13 @@ public function load(ObjectManager $manager)
             $integration = new Integration();
             $integration->setName(sprintf('Integration name %s', $i));
 
+            $username = sprintf('integration_%s@example.com', $i);
+            $user = new User();
+            $user->setEmail($username);
+            $user->setUsername($username);
```
Comment (kallosz) [inline src/SevenTag/Api/SecurityBundle/DataFixtures/ORM/IntegrationFixture.php]: this setter is not needed.

File: src/SevenTag/Api/SecurityBundle/Form/Type/IntegrationUserType.php [hunk 2]
```diff
@@ -43,7 +43,8 @@ public function buildForm(FormBuilderInterface $builder, array $options)
     public function configureOptions(OptionsResolver $resolver)
     {
         $resolver->setDefaults([
-            'data_class' => 'SevenTag\Api\UserBundle\Entity\User'
+            'data_class' => 'SevenTag\Api\UserBundle\Entity\User',
+            //'allow_extra_fields' => true
```
Comment (kallosz) [inline src/SevenTag/Api/SecurityBundle/Form/Type/IntegrationUserType.php]: remove this comment

File: src/SevenTag/Api/SecurityBundle/Resources/config/services.yml [hunk 3]
```diff
@@ -59,3 +62,11 @@ services:
...
         calls:
           - [setUserChecker, [@security.user_checker]]
+
+    seven_tag_security.security.oauth_token_user_resolver:
+        class: SevenTag\Api\SecurityBundle\Security\Firewall\OAuthTokenUserResolver
+        arguments: [@fos_oauth_server.access_token_manager.default]
+
+    fos_oauth_server.security.authentication.listener:
+        class: SevenTag\Api\SecurityBundle\Security\Firewall\OAuthListener
+        arguments: [@security.context, @security.authentication.manager, @fos_oauth_server.server, @seven_tag_security.security.oauth_token_user_resolver]
```
Comment (kallosz) [inline src/SevenTag/Api/SecurityBundle/Resources/config/services.yml]: too long line. change format to:
```
arguments: 
   - @.... 
   - @..
```

File: src/SevenTag/Api/SecurityBundle/Security/Firewall/OAuthListener.php [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+    }
+
+    /**
+     * @param GetResponseEvent $event The event.
+     */
+    public function handle(GetResponseEvent $event)
+    {
+        parent::handle($event);
+
+        if ($token = $this->securityContext->getToken()) {
```
Comment (kallosz) [inline src/SevenTag/Api/SecurityBundle/Security/Firewall/OAuthListener.php]: reverse this condition to decrease complexity.

File: src/SevenTag/Api/SecurityBundle/Security/Firewall/OAuthTokenUserResolver.php [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+     * @param UserInterface $user
+     * @return OAuthToken
+     */
+    protected function createToken(Token $accessToken, UserInterface $user)
+    {
+        $scope = $accessToken->getScope();
+        $roles = $user->getRoles();
+
+        if (!empty($scope)) {
+            foreach (explode(' ', $scope) as $role) {
```
Comment (kallosz) [inline src/SevenTag/Api/SecurityBundle/Security/Firewall/OAuthTokenUserResolver.php]: `explode(' ', $scope)` move this to separate method.

File: src/SevenTag/Api/SecurityBundle/Security/Firewall/OAuthTokenUserResolver.php [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+                $roles[] = 'ROLE_' . strtoupper($role);
+            }
+        }
+
+        $token = new OAuthToken($roles);
```
Comment (kallosz) [inline src/SevenTag/Api/SecurityBundle/Security/Firewall/OAuthTokenUserResolver.php]: create named constructor for this to increase readability.

File: src/SevenTag/Api/SecurityBundle/Security/Firewall/OAuthTokenUserResolver.php [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+     * @return OAuthToken
+     */
+    protected function createToken(Token $accessToken, UserInterface $user)
+    {
+        $scope = $accessToken->getScope();
+        $roles = $user->getRoles();
+
+        if (!empty($scope)) {
+            foreach (explode(' ', $scope) as $role) {
+                $roles[] = 'ROLE_' . strtoupper($role);
```
Comment (kallosz) [inline src/SevenTag/Api/SecurityBundle/Security/Firewall/OAuthTokenUserResolver.php]: `'ROLE_' . strtoupper($role);` move this to separate method.

File: src/SevenTag/Api/SecurityBundle/Security/Firewall/OAuthTokenUserResolver.php [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+
+    /**
+     * @param OAuthToken $token
+     * @return OAuthToken
+     */
+    protected function resolveAccessTokenUser(OAuthToken $token)
+    {
+        $accessToken = $this->tokenManager->findTokenByToken($token->getToken());
+
+        if ($accessToken instanceof Token && is_null($accessToken->getUser())) {
```
Comment (kallosz) [inline src/SevenTag/Api/SecurityBundle/Security/Firewall/OAuthTokenUserResolver.php]: move this to separate method to increase readability and reverse condition to decrease complexity.

File: src/SevenTag/Api/SecurityBundle/Tests/Security/Firewall/OAuthTokenUserResolverTest.php [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+    public function itShouldReturnTokenWithUser()
+    {
+        $token = new OAuthToken();
+        $token->setToken('test');
+
+        $user = new User();
+        $accessToken = new AccessToken();
+        $client = new Integration();
+        $client->setUser($user);
+        $accessToken->setClient($client);
```
Comment (kallosz) [inline src/SevenTag/Api/SecurityBundle/Tests/Security/Firewall/OAuthTokenUserResolverTest.php]: move this whole block to separate method. we don't need to know when we read test how this is created. 
Name this properly and create new method for this.

File: src/SevenTag/Api/UserBundle/Entity/User.php [hunk 10]
```diff
@@ -195,4 +195,16 @@ public function getDisplayName()
 
         return $this->getEmail();
     }
+
+    /**
+     * @param string $email
+     * @return $this
```
Comment (kallosz) [inline src/SevenTag/Api/UserBundle/Entity/User.php]: should be `return ClassName`

File: src/SevenTag/Api/SecurityBundle/Form/Type/EditIntegrationType.php [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+{
+    /**
+     * {@inheritdoc}
+     */
+    public function configureOptions(OptionsResolver $resolver)
+    {
+        $resolver->setDefaults([
+            'data_class' => 'SevenTag\Api\SecurityBundle\Entity\Integration',
+            'validation_groups' => ['EditIntegration'],
+            'allow_extra_fields' => true
```
Comment (kallosz) [inline src/SevenTag/Api/SecurityBundle/Form/Type/EditIntegrationType.php]: this is wrong. we can't allow for extra fileds.

File: src/SevenTag/Api/SecurityBundle/Form/Type/CreateIntegrationType.php [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+{
+    /**
+     * {@inheritdoc}
+     */
+    public function buildForm(FormBuilderInterface $builder, array $options)
+    {
+        parent::buildForm($builder, $options);
+
+        $builder
+            ->add('user', new IntegrationUserType());
```
Comment (kallosz) [inline src/SevenTag/Api/SecurityBundle/Form/Type/CreateIntegrationType.php]: missing `name` field?

File: src/SevenTag/Api/SecurityBundle/Form/Type/EditIntegrationType.php [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+
+use Symfony\Component\OptionsResolver\OptionsResolver;
+
+/**
+ * Class EditIntegrationType
+ * @package SevenTag\Api\SecurityBundle\Form\Type
+ */
+class EditIntegrationType extends IntegrationType
+{
+    /**
```
Comment (kallosz) [inline src/SevenTag/Api/SecurityBundle/Form/Type/EditIntegrationType.php]: edit integration should have removed user field. We can't change already assigned user.

File: src/SevenTag/Api/SecurityBundle/Security/Firewall/OAuthTokenUserResolver.php (hunk 7 above)
Comment (michalsikora) [inline src/SevenTag/Api/SecurityBundle/Security/Firewall/OAuthTokenUserResolver.php]: agree with move to separate method. Do not agree with created class named `Utils` this is to much common name without basic information about bussiness logic.

File: src/SevenTag/Api/SecurityBundle/Tests/Security/Firewall/OAuthListenerTest.php [hunk 14]
```diff
@@ -0,0 +1,120 @@
...
+use SevenTag\Api\UserBundle\Entity\User;
+use Symfony\Component\Security\Core\Authentication\Token\Storage\TokenStorage;
+
+/**
+ * Class OAuthListenerTest
+ * @package SevenTag\Api\SecurityBundle\Tests\Security\Firewall
+ */
+class OAuthListenerTest extends \PHPUnit_Framework_TestCase
+{
+    protected $securityContext;
```
Comment (kallosz) [inline src/SevenTag/Api/SecurityBundle/Tests/Security/Firewall/OAuthListenerTest.php]: missing docs

File: src/SevenTag/Api/SecurityBundle/Tests/Security/Firewall/OAuthTokenUserResolverTest.php [hunk 15]
```diff
@@ -0,0 +1,156 @@
...
+use SevenTag\Api\UserBundle\Entity\User;
+use Symfony\Component\Security\Core\Authentication\Token\Storage\TokenStorage;
+
+/**
+ * Class OAuthTokenUserResolverTest
+ * @package SevenTag\Api\SecurityBundle\Tests\Security\Firewall
+ */
+class OAuthTokenUserResolverTest extends \PHPUnit_Framework_TestCase
+{
+    protected $tokenManager;
```
Comment (kallosz) [inline src/SevenTag/Api/SecurityBundle/Tests/Security/Firewall/OAuthTokenUserResolverTest.php]: missing docs

File: src/SevenTag/Api/SecurityBundle/Security/Firewall/OAuthTokenUserResolver.php (hunk 7 above)
Comment (kallosz) [inline src/SevenTag/Api/SecurityBundle/Security/Firewall/OAuthTokenUserResolver.php]: @michalsikora I also agree that class `Utils` is too common. But this is used in serval places and we should find common space for this.

**Pull request merged**